compiler/rustc_typeck/src/check/callee.rs

```diff
@@ -374,7 +374,29 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                     |p| format!("`{}` defined here returns `{}`", p, callee_ty),
                                 )
                             }
-                            _ => Some(format!("`{}` defined here", callee_ty)),
+                            _ => {
+                                match def {
+                                    // Emit a different diagnostic for local variables, as they are not
+                                    // type definitions themselves, but rather variables *of* that type.
+                                    Res::Local(hir_id) => Some(format!(
+                                        "`{}` has type `{}`",
+                                        self.tcx.hir().name(hir_id),
+                                        callee_ty
+                                    )),
+                                    Res::Def(kind, def_id) => {
+                                        if let Some(Namespace::ValueNS) = kind.ns() {
+                                            Some(format!(
+                                                "`{}` has type `{}`",
+                                                self.tcx.def_path_str(def_id),
+                                                callee_ty
+                                            ))
```

Comment:
> @camelid I did the check that you have suggested but it does bring us back to the original problem, like this:
> 
> ```
> LL | struct Empty2;
>    | -------------- is of type `Empty2`
> ```
> 
> It should say `Empty2 defined here`. Is there a better way to do this?

I think it's fine the way it is now—in fact, I think it's better than the `Empty2 defined here` that I originally suggested—but if you don't want to show the type, I think you can just do

```suggestion
                                            Some(format!(
                                                "`{}` defined here",
                                                self.tcx.def_path_str(def_id),
                                            ))
```